### ISSUE TITLE
Camera Roll calculation

### ISSUE THREAD

**Issue by SlySnake** [AnalyticalGraphicsInc/cesium#2446]
I've tested the roll calculation in camera.js of the current source (v1.6). It seems to only use the z of the right vector. That works fine with level pitch 0. However as pitch deviates from 0 the more roll is off, because the z of right will then never go all the way to 1 which acos expects. For example: with a pitch of -45 if you roll on your left side roll value is only 45 when it should be 90.

Rather than:
var roll = Math.acos(-right.z) - CesiumMath.PI_OVER_TWO;

This seems to work for any pitch value except +90 and -90 (uses up and right vectors):
var roll = Math.atan2(right.z,up.z);

I've used this formula in 6DOF Google Earth as (heading/tilt/roll) is the only way to communicate orientation in Google Earth API. I'm already using it in Cesium in a readout and it works fine as well, using the camera's ENU vectors.

Pitch +90 and Pitch -90 will both yield Math.atan2(0,0). At these pitch values perhaps:
-heading could be NA and heading of the up vector could be roll
-or roll could NA and heading of the up vector could be heading
(maybe down vector for pitch +90, up vector for pitch -90)

I've also tested the older roll calculation
Math.atan2(right.z, right.x);
and had these results:

Pitch -45, Heading 0
Works fine

Pitch -45, Heading 90
Roll on left side yields +45 roll, should be +90
Roll on right side yields -135 roll, should be -90

Pitch 0, Heading 90
constant roll speed: You'll notice roll value changes very fast near 0, but changes very slow near 90, when it's supposed to change at a constant rate everywhere.

**Comment by pjcozzi** [comment]
Thanks @SlySnake.  @bagnell what do you think?

**Comment by SlySnake** [comment]
I'm not sure if get roll is used at the moment. Although back in September
https://github.com/AnalyticalGraphicsInc/cesium/commit/41e65419aeb4677f7b5e814f260ce099a2120658 line 773
I did notice that set roll used 'current roll' and 'desired roll' to obtain the 'delta roll' for this.look(this.direction, angle); Without having the proper current roll this wouldn't work properly. Perhaps set roll could now be resurrected?

I made a temporary unlisted group of videos to demonstrate roll at 4 different heading pitch combos
https://www.youtube.com/playlist?list=PLu7ubzox5Ny4HIOEgOWaw3Osn4blZXYEp
(be sure to use 1080p60 to read the small text)
(the acos calculation also uses the if (right.z < 0.0) modification)

Truth be told I haven't yet figured out why atan2 right z up z works. When one thinks of atan2 they think of y and x components of one vector to determine direction from origin in 2 dimensions. But z of one and z of another vector is odd. The fact that they are perpendicular vectors probably has something to do with it. Unlike acos, atan2 doesn't compare to 1, rather it compares both parameters to eachother it seems.
